plugin/evm/warp/backend.go

```diff
@@ -0,0 +1,76 @@
+// (c) 2019-2020, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package warp
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/cache"
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow"
+	"github.com/ava-labs/avalanchego/utils/hashing"
+	"github.com/ava-labs/avalanchego/vms/platformvm/teleporter"
+)
+
+var _ WarpBackend = &warpBackend{}
+
+// WarpBackend tracks signature eligible warp messages and provides an interface to fetch them.
+// The backend is also used to query for warp message signatures by the signature request handler.
+type WarpBackend interface {
+	// AddMessage signs [unsignedMessage] and adds it to the warp backend database
+	AddMessage(ctx context.Context, unsignedMessage *teleporter.UnsignedMessage) error
+
+	// GetSignature returns the signature of the requested message hash.
+	GetSignature(ctx context.Context, messageHash ids.ID) ([]byte, error)
+}
+
+// warpBackend implements WarpBackend, keeps track of warp messages, and generates message signatures.
+type warpBackend struct {
+	db             database.Database
+	snowCtx        *snow.Context
+	signatureCache *cache.LRU
+}
+
+// NewWarpBackend creates a new WarpBackend, and initializes the signature cache and message tracking database.
+func NewWarpBackend(snowCtx *snow.Context, db database.Database, signatureCacheSize int) WarpBackend {
+	return &warpBackend{
+		db:             db,
+		snowCtx:        snowCtx,
+		signatureCache: &cache.LRU{Size: signatureCacheSize},
+	}
+}
+
+func (w *warpBackend) AddMessage(ctx context.Context, unsignedMessage *teleporter.UnsignedMessage) error {
+	messageID := hashing.ComputeHash256Array(unsignedMessage.Bytes())
+
+	// We generate the signature here and only save the signature in the db and cache.
+	// It is left to smart contracts built on top of Warp to save messages if required.
+	signature, err := w.snowCtx.TeleporterSigner.Sign(unsignedMessage)
+	if err != nil {
+		return fmt.Errorf("failed to sign warp message: %w", err)
+	}
+
+	if err := w.db.Put(messageID[:], signature); err != nil {
+		return fmt.Errorf("failed to put warp signature in db: %w", err)
+	}
+
+	w.signatureCache.Put(ids.ID(messageID), signature)
+	return nil
+}
+
+func (w *warpBackend) GetSignature(ctx context.Context, messageID ids.ID) ([]byte, error) {
+	if sig, ok := w.signatureCache.Get(messageID); ok {
+		return sig.([]byte), nil
+	}
+
+	signature, err := w.db.Get(messageID[:])
```

Comment:
Was thinking when this case happens, since if we have the messageID in the db, the signature is likely in the cache as well. I think it happens when the cache size starts getting full and we evict signatures from the cache, then we can call the db for the old signatures.

Reply:
That's true, I'd expect it to be more likely that a recently added message will need to be accessed from cache, so might be better to not add an old entry back into the cache.

Reply:
The more common case (compared to cache eviction) to consider here is the restart of the node. The cache will be empty, but it is likely for a relayer to query for signatures of recent messages still, and we'd still want to cache those results rather than only caching signatures on initial insertion into the DB. So I think it's fine to insert it back into the cache here still :+1:

Reply:
That makes sense to me.

One possible thing to consider: As it stands, a malicious relayer could purposely query for old entries as a way to pollute the cache. There is a small amplification factor to this attack since a node has to not only serve the request from the DB, but may also have to serve a later request from the DB (that otherwise would have been served from the cache) due to a cache miss.

I don't think we need to worry about this, and the solution is to rate limit requests from each relayer rather than to change our caching approach.